lisa/lisa-analyses/src/main/java/it/unive/lisa/analysis/heap/pointbased/FieldSensitivePointBasedHeap.java

```diff
@@ -44,7 +56,77 @@ public FieldSensitivePointBasedHeap() {
 	 * @param heapEnv the heap environment that this instance tracks
 	 */
 	public FieldSensitivePointBasedHeap(HeapEnvironment<AllocationSites> heapEnv) {
+		this(heapEnv, new HashMap<>());
+	}
+
+	/**
+	 * Builds a new instance of field-sensitive point-based heap from its heap
+	 * environment.
+	 * 
+	 * @param heapEnv the heap environment that this instance tracks
+	 * @param fields  the mapping between allocation sites and their fields that
+	 *                    this instance tracks
+	 */
+	public FieldSensitivePointBasedHeap(HeapEnvironment<AllocationSites> heapEnv,
+			Map<AllocationSite, Set<SymbolicExpression>> fields) {
 		super(heapEnv);
+		this.fields = fields;
+	}
+
+	@Override
+	public PointBasedHeap assign(Identifier id, SymbolicExpression expression, ProgramPoint pp)
+			throws SemanticException {
+
+		PointBasedHeap sss = smallStepSemantics(expression, pp);
+		ExpressionSet<ValueExpression> rewrittenExp = sss.rewrite(expression, pp);
+
+		PointBasedHeap result = bottom();
+		for (ValueExpression exp : rewrittenExp)
+			if (exp instanceof MemoryPointer) {
+				MemoryPointer pid = (MemoryPointer) exp;
+				HeapLocation star_y = pid.getReferencedLocation();
+				if (id instanceof MemoryPointer) {
+					// we have x = y, where both are pointers
+					// we perform *x = *y so that x and y
+					// become aliases
+					Identifier star_x = ((MemoryPointer) id).getReferencedLocation();
+					HeapEnvironment<AllocationSites> heap = sss.heapEnv.assign(star_x, star_y, pp);
+					result = result.lub(from(new PointBasedHeap(heap)));
+				} else {
+					if (star_y instanceof StaticAllocationSite) {
+						// no aliasing: star_y must be cloned and the clone must
+						// be assigned to id
+						StaticAllocationSite clone = new StaticAllocationSite(star_y.getStaticType(),
+								id.getCodeLocation().toString(), star_y.isWeak(), id.getCodeLocation());
+						HeapEnvironment<AllocationSites> heap = sss.heapEnv.assign(id, clone, pp);
+						result = result.lub(from(new PointBasedHeap(heap)));
+
+						// all the allocation sites fields of star_y
+						if (fields.containsKey(star_y))
```

Comment:
This whole method seems the same as the assign method for `PointBasedHeap`, except for this case. We should have a unique implementation with protected methods that are overridden when needed.